### DISCUSSION TITLE
Simplify road network by adding nodes at turns in edges and drawing straight line between the nodes

### DISCUSSION THREAD

**Discussion by timverlaan** [luukvdmeer/sfnetworks#222] [Q&A] [answered]
Hi all,

I am using road network data from OSM in an ABM made in NetLogo. I am having trouble with agents getting stuck in the network and would therefore like to simplify the network. I would like to simplify it by adding nodes at turns in edges and drawing straight line between the nodes. This will make my turn less smooth, but this is manageable. Below some code to illustrate and a image of what I mean to do. I have spoken to Luuk and understand this is not yet possible in _sfnetwork_, however, I was hoping someone here might know a solution to my problem... 

Thanks in advance!

```r
# get Amsterdam centre boxed in
area <- getbb("Amsterdam")

#load streets from OSM
streets <- area %>%
  opq() %>%
    add_osm_feature(key = "highway", value = c("motorway",
                    "trunk", "primary", "secondary",
                    "tertiary", "unclassified", "residential",
                    "motorway_link", "trunk_link", "primary_link",
                    "secondary_link", "tertiary_link", "living_street", "road", "unclassified",
                    "pedestrian")
                  ) %>%
  osmdata_sf(quiet = FALSE) 

streets_4326 <- streets$osm_lines

# make osm file into sfnetwork file
my_sfn <- as_sfnetwork(streets_4326, directed = TRUE, length_as_weight = TRUE)

tm_shape(smoothed %>% activate(edges) %>% as_tibble() %>% st_as_sf()) +
  tm_lines() +
  tm_shape(smoothed %>% activate(nodes) %>% as_tibble() %>% st_as_sf()) +
  tm_dots() +
  tmap_options(basemaps = 'OpenStreetMap')

```

![image](https://user-images.githubusercontent.com/26186975/185159634-95456d6d-c0c0-4025-8c42-9e4407b7f8db.png)

should be this: 

![image](https://user-images.githubusercontent.com/26186975/185160666-80e82447-53a8-408c-8dca-e8a614514d8a.png)

(I know I still have to use to_spatial_subdivision here first to make a node at "Gedempte Hamerkanaal")

**Comment by agila5** [comment]
Hi! I think that what you are proposing might be doable after implementing https://github.com/luukvdmeer/sfnetworks/issues/177. 

In particular, I think you could adopt the following approach: 

1. Divide the edge at each internal point using the morpher described here in https://github.com/luukvdmeer/sfnetworks/pull/210 since it might be difficult to work with POINT(s) that are not nodes; 
2. Define the meaning of "nodes at turns in edges turn in edges" (i.e. maybe something like "the intersecting points between two street segments where the angle at the intersection is greater than some threshold) and calculate those points; 
3. Smooth all the other nodes while preserving the nodes at turns. This last step requires #177. 

The main problem here is that we are assuming that there exists an OSM node at each turn but, looking at the figures online, I think this is a reasonable assumption. What do you think?

**Reply by timverlaan** [reply]
I think point 2. is difficult as the turns are constructed by multiple small straight lines together constituting a smooth turn. In this sense, there is no consecutive pair of lines at a high enough angle to set as a threshold. However, maybe we can extrapolate the longest line segments and see where they would intersect, and put a node there?

_The main problem here is that we are assuming that there exists an OSM node at each turn but, looking at the figures online, I think this is a reasonable assumption. What do you think?_ 

If I look at left bottom (nodes) [here ](https://www.openstreetmap.org/way/310252575#map=19/52.38312/4.92147)I think that assumption is correct

**Reply by timverlaan** [reply]
[This ](https://github.com/osmlab/atlas/tree/dev/src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/bignode#bignode)Java package seems to do what I intend to. [This ](https://stackoverflow.com/questions/64644887/download-roads-network-from-openstreetmap-and-simplify-graph-of-roads)question on stack is facing the same problem.

**Comment by timverlaan** [comment, marked as answer]
I ended up the using the Ramer-Douglas-Peucker algorithm to do this!
